data/transactions/logic/opcodes.go

```diff
@@ -92,13 +101,38 @@ func stacky(typer opTypeFunc, imms ...string) opDetails {
 	return d
 }
 
-func varies(checker opCheckFunc, name string, kind immKind) opDetails {
-	return opDetails{1, 0, checker, []immediate{{name, kind}}, nil}
+func sizeVaries(checker opCheckFunc, name string, kind immKind) opDetails {
+	return opDetails{1, nil, 0, checker, []immediate{{name, kind, nil}}, nil}
+}
+
+// field is used to create an opDetails for an opcode with a single field
+func field(immediate string, group *FieldGroup) opDetails {
+	opd := immediates(immediate)
+	opd.Immediates[0].Group = group
+	return opd
+}
+
+// field is used to annotate an existing immediate with group info
+func (d opDetails) field(name string, group *FieldGroup) opDetails {
+	for i := range d.Immediates {
+		if d.Immediates[i].Name == name {
+			d.Immediates[i].Group = group
+			return d
+		}
+	}
+	panic(name)
 }
 
-func costlyImm(cost int, names ...string) opDetails {
-	opd := immediates(names...)
-	opd.Cost = cost
+func costByField(immediate string, group *FieldGroup, costs map[byte]int) opDetails {
+	opd := immediates(immediate).costs(0)
+	opd.Immediates[0].Group = group
+	opd.costFunc = func(program []byte, pc int) int {
+		cost, ok := costs[program[pc+1]]
+		if ok {
+			return cost
+		}
+		return 1
```

Comment:
@jannotti For my edification: What tradeoffs were considered when defaulting to cost = 1 here?

I understand `!ok` case must be handled.  Though I can imagine a couple of other recourses:
* Each opcode supplies a fallback default.
* AVM execution is interrupted.

I think there's implications I don't totally grok yet, which is part of the reason for raising the question.  In the abstract, my concern with _always_ returning 1 is that it _may_ result in execution costing less than it should.

Interrupting AVM execution feels abrasive.  However, on the surface, the condition seems exceptional.

Reply:
I'm deathly afraid that some opcode with have a cost of 0, and could somehow be used to hang the AVM.

As you say, we could instead panic if a 0 is returned.  As I think about it, I'm leaning that way.  It would be easy to convince me.

Reply:
I'm definitely going to change to failing.  Here's my reasoning.

If we have a bug, and 1 is returned, we'll end up with transactions on the blockchain that enshrine that bug forever.  If we have a bug where 0 is returned, but we also panic if the final cost is 0 for _any_ opcode, then we ensure that we never have a transaction succeed that uses the buggy behavior, and we can fix it.

Reply:
@jannotti The rationale makes sense to me and I'm mostly on-board with  https://github.com/algorand/go-algorand/pull/3832/commits/13b24999a6930ccdfdd35b2cf126df9e023a18d1.

My personal preference is to avoid relying on 0 as a sentinel value and instead return error.  However, I suspect there's a rationale for using 0.

Is the following reasoning accurate?
* An uninitialized struct produces 0.  
* So, even if an error is returned in some cases (e.g. `costFunc`), I think you're imagining that we always also need to check for 0.

If these statements are accurate, then I'll give it all some more thought out-of-band.  And I'll be good to resolve the thread.

Reply:
Yes, the deal is that we always must check 0, because it would be a disaster. So it's less about making it a sentinel and more about taking advantage of a check that ought to be there anyway.

Reply:
@jannotti Got it - My concerns here are addressed. Resolving.